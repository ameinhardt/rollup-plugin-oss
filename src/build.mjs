/* eslint no-console: ['error', { allow: ['info'] }] */
import { readdir, readFile, writeFile } from 'node:fs/promises';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import nearley from 'nearley';
import Compile from 'nearley/lib/compile.js';
import generate from 'nearley/lib/generate.js';
import nearleyGrammar from 'nearley/lib/nearley-language-bootstrapped.js';

const srcDir = dirname(fileURLToPath(import.meta.url));

async function buildLicenseDatabase() {
  const basedir = '../license-list-data/json',
    outfile = '../spdx.json',
    [licenseFiles, exceptionFiles] = await Promise.all([
      readdir(resolve(srcDir, basedir, 'details')),
      readdir(resolve(srcDir, basedir, 'exceptions'))
    ]),
    licenses = {},
    exceptions = {};
  console.info(`database: reading ${licenseFiles.length} licenses`);
  for (const filename of licenseFiles) {
    const data = await readFile(resolve(srcDir, basedir, 'details', filename)),
      detail = JSON.parse(data);
    licenses[detail.licenseId] = detail.licenseText.normalize('NFD').replace(/\s/g, ' ');
  }
  console.info(`database: reading ${exceptionFiles.length} exceptions`);
  for (const filename of exceptionFiles) {
    const data = await readFile(resolve(srcDir, basedir, 'exceptions', filename)),
      exception = JSON.parse(data);
    exceptions[exception.licenseExceptionId] = exception.licenseExceptionText.normalize('NFD').replace(/\s/g, ' ');
  }
  console.info(`database: writing ${outfile}`);
  await writeFile(resolve(srcDir, outfile), JSON.stringify({
    licenses,
    exceptions
  }, null, 2));
}

async function buildLicenseParser() {
  const outfile = './spdxExpression.ts',
    parserGrammar = nearley.Grammar.fromCompiled(nearleyGrammar),
    parser = new nearley.Parser(parserGrammar),
    grammar = await readFile(resolve(srcDir, 'spdxExpression.ne'));
  parser.feed(grammar.toString());
  const c = Compile(parser.results[0], {});
  console.info(`parser:   writing ${outfile}`);
  writeFile(resolve(srcDir, outfile), generate(c));
}

// eslint-disable-next-line unicorn/prefer-top-level-await
buildLicenseDatabase();
// eslint-disable-next-line unicorn/prefer-top-level-await
buildLicenseParser();
