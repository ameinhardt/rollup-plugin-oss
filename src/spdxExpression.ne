@preprocessor typescript

# https://spdx.dev/spdx-specification-21-web-version/#h.jxpfx0ykyb60

MAIN -> COMPOUNDEXPRESSION {% ([expression]) => expression %}
IDSTRING -> ([a-zA-Z]
			| [0-9]
			| "-"
			| "." ):+ {% ([chars]) => chars.join('') %}
LICENSEID -> IDSTRING {% id %}
LICENSEEXCEPTIONID -> IDSTRING {% id %}
LICENSEREF -> ("DocumentRef-" IDSTRING ":"):? "LicenseRef-" IDSTRING
SIMPLEEXPRESSION -> LICENSEID {% ([license]) => ({ license }) %}
					| LICENSEID "+" {% ([license]) => ({ license, plus: true }) %}
					| LICENSEREF {% ([license]) => ({ license }) %}

COMPOUNDEXPRESSION -> SIMPLEEXPRESSION {% ([expression]) => expression %}
					  | SIMPLEEXPRESSION __ "WITH" __ LICENSEEXCEPTIONID {% ([license,, conjuction,, exception]) => ({ ...license, exception}) %}
					  | COMPOUNDEXPRESSION __ ("AND" | "OR") __ COMPOUNDEXPRESSION {% ([left,, [conjunction],, right]) => ({ left, conjunction, right }) %}
					  | "(" _ COMPOUNDEXPRESSION _ ")" {% (expression) => expression[2] %}

_ -> " ":* {% () => null %}
__ -> " ":+ {% () => null %}
