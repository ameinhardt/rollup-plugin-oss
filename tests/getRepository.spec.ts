import { getRepository } from '../src/index';

// https://docs.npmjs.com/cli/v8/configuring-npm/package-json#repository
const possibleValues = [
  'npm/npm',
  'github:user/repo',
  'gist:11081aaa281',
  'bitbucket:user/repo',
  'gitlab:user/repo',
  42,
  {
    type: 'git',
    url: 'https://github.com/npm/cli.git'
  },
  {
    type: 'git',
    url: 'https://github.com/facebook/react.git',
    directory: 'packages/react-dom'
  }];

describe('gets a respository url from package attribute', function() {
  for (const [index, value] of possibleValues.entries()) {
    it(`#${index + 1} ${typeof value === 'string' ? value : 'object' }`, function() {
      try {
        const repository = getRepository(value as string);
        expect(typeof repository).toBe('string');
        expect(repository).toMatchSnapshot();
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    });
  }
});
