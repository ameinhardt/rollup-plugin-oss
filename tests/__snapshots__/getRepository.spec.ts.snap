// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets a respository url from package attribute #1 npm/npm 1`] = `"https://github.com/npm/npm"`;

exports[`gets a respository url from package attribute #2 github:user/repo 1`] = `"https://github.com/user/repo"`;

exports[`gets a respository url from package attribute #3 gist:11081aaa281 1`] = `"https://gist.github.com/11081aaa281"`;

exports[`gets a respository url from package attribute #4 bitbucket:user/repo 1`] = `"https://bitbucket.org/user/repo"`;

exports[`gets a respository url from package attribute #5 gitlab:user/repo 1`] = `"https://gitlab.com/user/repo"`;

exports[`gets a respository url from package attribute #6 object 1`] = `""`;

exports[`gets a respository url from package attribute #7 object 1`] = `"https://github.com/npm/cli"`;

exports[`gets a respository url from package attribute #8 object 1`] = `"https://github.com/facebook/react"`;
